pkg/cluster/loadbalancerprofile.go

```diff
@@ -0,0 +1,396 @@
+	desiredCount := 0
+	if defaultIP, ok := ipAddresses[infraID+"-pip-v4"]; ok {
+		desiredIPAddresses = append(desiredIPAddresses, api.ResourceReference{ID: *defaultIP.ID})
+		desiredCount += 1
+		delete(ipAddresses, infraID+"-pip-v4")
+	}
+
+	for _, v := range ipAddresses {
+		if desiredCount < managedOBIPCount {
+			desiredIPAddresses = append(desiredIPAddresses, api.ResourceReference{ID: *v.ID})
+			desiredCount += 1
+		} else {
+			break
+		}
+	}
+
+	return desiredIPAddresses, nil
+}
+
+// Get all current managed IP Addresses in cluster resource group based on naming convention.
+func (m *manager) getClusterManagedIPs(ctx context.Context) (map[string]mgmtnetwork.PublicIPAddress, error) {
+	resourceGroupName := stringutils.LastTokenByte(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+	infraID := m.doc.OpenShiftCluster.Properties.InfraID
+	var ipAddresses = make(map[string]mgmtnetwork.PublicIPAddress)
+
+	result, err := m.publicIPAddresses.List(ctx, resourceGroupName)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < len(result); i++ {
+		// <infraID>-pip-v4 is not necessarily managed but is the default installed outbound IP
+		if *result[i].Name == infraID+"-pip-v4" || strings.Contains(*result[i].Name, "-outbound-pip-v4") {
+			ipAddresses[*result[i].Name] = result[i]
+		}
+	}
+
+	return ipAddresses, err
+}
+
+func genManagedOutboundIPName() string {
+	return uuid.DefaultGenerator.Generate() + "-outbound-pip-v4"
+}
+
+// Create a managed outbound IP Address.
+func (m *manager) createPublicIPAddress(ctx context.Context) (mgmtnetwork.PublicIPAddress, error) {
+	name := genManagedOutboundIPName()
+	resourceGroupName := stringutils.LastTokenByte(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+	resourceID := fmt.Sprintf("%s/providers/Microsoft.Network/publicIPAddresses/%s", m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, name)
+	m.log.Infof("creating public IP Address: %s", name)
+	publicIPAddress := mgmtnetwork.PublicIPAddress{
+		Name:     &name,
+		ID:       &resourceID,
+		Location: &m.doc.OpenShiftCluster.Location,
+		PublicIPAddressPropertiesFormat: &mgmtnetwork.PublicIPAddressPropertiesFormat{
+			PublicIPAllocationMethod: mgmtnetwork.Static,
+			PublicIPAddressVersion:   mgmtnetwork.IPv4,
+		},
+		Sku: &mgmtnetwork.PublicIPAddressSku{
+			Name: mgmtnetwork.PublicIPAddressSkuNameStandard,
+		},
+	}
+
+	err := m.publicIPAddresses.CreateOrUpdateAndWait(ctx, resourceGroupName, name, publicIPAddress)
+	if err != nil {
+		return mgmtnetwork.PublicIPAddress{}, err
+	}
+
+	return publicIPAddress, nil
+}
+
+func getOutboundIPsFromLB(lb mgmtnetwork.LoadBalancer) []api.ResourceReference {
+	var outboundIPs []api.ResourceReference
+	fipConfigs := make(map[string]mgmtnetwork.FrontendIPConfiguration, len(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations))
+
+	for i := 0; i < len(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations); i++ {
```

Comment:
Could we get a nil pointer deref?

Reply:
There should always be data in  `*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations`.  There should  be at least 1 FrontendIPConfiguration containing the IP used by the outbound rule `outbound-rule-v4`.